bindings/pyroot_experimental/PyROOT/src/TClonesArrayPyz.cxx

```diff
@@ -0,0 +1,125 @@
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+// Bindings
+#include "CPyCppyy.h"
+#include "PyROOTPythonize.h"
+#include "CPPInstance.h"
+#include "MemoryRegulator.h"
+#include "Utility.h"
+#include "PyzCppHelpers.hxx"
+
+// ROOT
+#include "TClass.h"
+#include "TClonesArray.h"
+
+using namespace CPyCppyy;
+
+// Helper: converts Python index into straight C index.
+static PyObject *PyStyleIndex(PyObject *self, PyObject *index)
+{
+   Py_ssize_t idx = PyInt_AsSsize_t(index);
+   if (idx == (Py_ssize_t)-1 && PyErr_Occurred())
+      return nullptr;
+
+   // To know the capacity of a TClonesArray, we need to invoke GetSize
+   PyObject *pysize = CallPyObjMethod(self, "GetSize");
+   if (!pysize) {
+      PyErr_Clear();
+      return nullptr;
+   }
+
+   Py_ssize_t size = PyInt_AsSsize_t(pysize);
+   Py_DECREF(pysize);
+   if (idx >= size || (idx < 0 && idx < -size)) {
+      PyErr_SetString(PyExc_IndexError, "index out of range");
+      return nullptr;
+   }
+
+   PyObject *pyindex = nullptr;
+   if (idx >= 0) {
+      Py_INCREF(index);
+      pyindex = index;
+   } else {
+      pyindex = PyLong_FromSsize_t(size + idx);
+   }
+
+   return pyindex;
+}
+
+// Customize item setting
+PyObject *SetItem(CPPInstance *self, PyObject *args)
+{
+   CPPInstance *pyobj = nullptr;
+   PyObject *idx = nullptr;
+   if (!PyArg_ParseTuple(args, const_cast<char *>("OO!:__setitem__"), &idx, &CPPInstance_Type, &pyobj))
+      return nullptr;
+
+   if (!self->GetObject()) {
+      PyErr_SetString(PyExc_TypeError, "unsubscriptable object");
+      return nullptr;
+   }
+
+   PyObject *pyindex = PyStyleIndex((PyObject *)self, idx);
+   if (!pyindex)
+      return nullptr;
+   int index = (int)PyLong_AsLong(pyindex);
+   Py_DECREF(pyindex);
+
+   // Get hold of the actual TClonesArray
+   auto cla = (TClonesArray *)GetTClass(self)->DynamicCast(TClonesArray::Class(), self->GetObject());
+
+   if (!cla) {
+      PyErr_SetString(PyExc_TypeError, "attempt to call with null object");
+      return nullptr;
+   }
+
+   if (Cppyy::GetScope(cla->GetClass()->GetName()) != pyobj->ObjectIsA()) {
+      PyErr_Format(PyExc_TypeError, "require object of type %s, but %s given", cla->GetClass()->GetName(),
+                   Cppyy::GetFinalName(pyobj->ObjectIsA()).c_str());
+   }
+
+   // Destroy old object, if applicable
+   if (((const TClonesArray &)*cla)[index]) {
+      cla->RemoveAt(index);
+   }
+
+   if (pyobj->GetObject()) {
+      // Accessing an entry will result in new, uninitialized memory (if properly used)
+      TObject *object = (*cla)[index];
+      pyobj->CppOwns();
+      MemoryRegulator::RegisterPyObject(pyobj, object);
+      memcpy((void *)object, pyobj->GetObject(), cla->GetClass()->Size());
```

Comment:
In general this is not a safe way of copying the objects. [It might work in many cases but many other will leave dangling pointers in the target].  See TObject::Copy which might be the solution here.

See also TClonesArray::ConstructedAt and TClonesArray::Absorb (i.e. or maybe we need a special case for pyroot added to TClonesArray).

Reply:
I'm trying to use `TObject::Copy` instead of `memcpy` but it seems to have no effect, the destination object remains null. I tried a quick test:
```
root [0] TObjString o("a");
root [1] TObjString o2("b");
root [2] cout << o.GetName() << endl;
a
root [3] cout << o2.GetName() << endl;
b
root [4] o.Copy(o2)
root [5] cout << o.GetName() << endl;
a
root [6] cout << o2.GetName() << endl;
b
```
I would expect `o` to be copied into `o2` but that does not seem the case. What I am missing? Thanks!

Reply:
that TObjString does not overload Copy .. :(  ... so indeed we can't rely on it ....

Reply:
The 'solution' is to use/implement a variant of TDirectory::CloneObject that replace the first statement (calling New()) with an input parameter ....

Reply:
Hi @pcanal, following your suggestion I modified the solution to use a helper method inspired on `TDirectory::CloneObject`, so that the cloning is done with streamers. Please review the corresponding commit b869970d14